### DISCUSSION TITLE
Bug Report: Allowed memory size exhausted while returning a collection datasource

### DISCUSSION THREAD

**Discussion by BenOussama180** [Power-Components/livewire-powergrid#574] [Q&A]
Here is a link to the error image : https://flareapp.io/share/NPG6p9nm#F1
Basically i want to remove duplicates from my Link model and from those duplicates i want to only retreive the cheapest one based on  price !
My datasource  : 
`public function datasource(): ?Collection
    {
        return Link::all()
            ->groupBy('site')
            ->map(function ($group) {
                $sorted = $group->sortBy('price');
                return $sorted->first();
            });
    }`
can i achieve the same thing witth a builder?

**Comment by luanfreitasdev** [comment]
Hello @BenOussama180 , I believe this is not a bug, because it is not possible to make a pagination using collection causing a lot of slowness when loading the table.

A suggestion would be to make a pagination outside the powergrid, using` take()` and `limit()`. So the manipulation would be outside and powergrid would take care of displaying only the requested data

**Reply by BenOussama180** [reply]
@luanfreitasdev ,i managed to make it work by using this : ini_set("memory_limit", "512M");
the page is indeed much slower thn it was, but i couldnt find a work around it, do you have any idea on how to transform my datasource into returning a builder ? an alternative of map() inside builder ?

**Reply by luanfreitasdev** [reply]
You don't need to call all records to display in a collection. This is going to be very slow, you can get 2 parameters inside the table $take and $limit to do Model::take($this->take)->limit($this->limit)->map(.....
